### DISCUSSION TITLE
Why does Utest not print summary report when run in mill build.sc ?

### DISCUSSION THREAD

**Discussion by jk-1** [com-lihaoyi/mill#1900] [Q&A]
I am porting an old `sbt` based project to mill. Project uses `uTest` for testing.

In sbt when I run `sbt test` the `sbt` runs the utests and after all tests `uTest` (or `sbt`) prints a nice summary on test success/failure.

When the exactly same code is run in `mill` `build.sc` with `mill gentool_2.test` command, this summary is not printed. How can I get this summary printed also with `mill`?

Used versions are:
```
Scala 2.12.9
Ammonite 2.2.0
uTest 0.7.11
mill 0.10.4
```

Simplified `build.sc` : 

```scala

val rootMillPath = millSourcePath

object gentool_2 extends ScalaModule {
  val utestV = "0.7.11"
  val ammoniteV = "2.2.0"
  def scalaV = "2.12.9"// scalaVer

  override def scalaVersion = scalaV
  override def ammoniteVersion = ammoniteV

  def ivyDeps = Agg(
    //...
  )

  val genToolSourceRoot =       s"${rootMillPath}/src/main/scala"
  val genToolTestRoot =         s"${rootMillPath}/src/test/scala"
  val genToolResourceRoot =     s"${rootMillPath}/src/main/resources"
  val genToolTestResourceRoot = s"${rootMillPath}/src/test/resources"

  override def sources: Sources = T.sources {
    Seq(PathRef(Path(genToolSourceRoot)))
  }

  override def resources: Sources = T.sources {
    Seq(PathRef(Path(genToolResourceRoot)))
  }

  object test extends ScalaModule with Tests with TestModule.Utest {
    override def scalaVersion = scalaV
    override def ammoniteVersion = ammoniteV

    def ivyDeps = gentool_2.ivyDeps() ++ Agg(
      ivy"com.lihaoyi::utest:${utestV}"
    )

    override def sources: Sources = T.sources {
      Seq(PathRef(Path(genToolTestRoot)))
    }

    override def resources: Sources = T.sources {
      Seq(PathRef(Path(genToolTestResourceRoot)))
    }
  }
}

```

**Comment by lefou** [comment]
I can't tell whether this is what you expected from sbt (as you have not provided any sbt test output to show what you mean with summary), but utest typically only prints a summary when the number of tests (failed tests) exceeds a threshold. According to the documentation, the default summary threshold is `20` and can also be customized (https://github.com/com-lihaoyi/utest#per-run-setupteardown-and-other-test-running-config).

As an additional note regarding your ported build script: Mill also has a convenience trait `SbtModule` which supports a sbt-like source directory layout. When using it, your `build.sc` from above can be simplified as follows:

```scala

object gentool_2 extends SbtModule {
  // use one level above as module base dir
  def millSourcePath = super.millSourcePath / os.up

  val utestV = "0.7.11"
  val ammoniteV = "2.2.0"
  def scalaV = "2.12.9"// scalaVer

  override def scalaVersion = scalaV
  override def ammoniteVersion = ammoniteV

  def ivyDeps = Agg(
    //...
  )

  object test extends Tests with TestModule.Utest {
    override def ammoniteVersion = ammoniteV

    def ivyDeps = Agg(
      ivy"com.lihaoyi::utest:${utestV}"
    )
  }
}


```
